File: python/tests/test_integration.py
"""
DevBuddy Integration Test — Week 0

Validates the full setup end-to-end:
1. .env is configured (OPENROUTER_API_KEY, DEVBUDDY_MODEL)
2. OpenRouter connectivity works
3. Structured output returns typed objects
4. Cost tracking works
5. Model swap works (optional)

Run: python tests/test_integration.py
"""
import os
import sys
import time

# Ensure src/ is importable (Python 3.12+ doesn't auto-add CWD)
# __file__ is python/tests/test_integration.py, so parent is python/
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from src.config import OPENROUTER_API_KEY, DEVBUDDY_MODEL, DEVBUDDY_MODEL_ALT

# ─── Test 1: Environment ──────────────────────────────────────
def test_env_configured():
    """Verify .env has the required variables."""
    assert OPENROUTER_API_KEY is not None, "❌ OPENROUTER_API_KEY not set in .env"
    assert "sk-or-" in OPENROUTER_API_KEY, "❌ OPENROUTER_API_KEY doesn't look like an OpenRouter key"
    assert DEVBUDDY_MODEL is not None, "❌ DEVBUDDY_MODEL not set in .env"

    print(f"  ✅ .env configured: model={DEVBUDDY_MODEL}, key={'*' * 8}{OPENROUTER_API_KEY[-4:]}")


# ─── Test 2: OpenRouter Connectivity ──────────────────────────
def test_openrouter_connectivity():
    """Verify we can call OpenRouter and get a response."""
    from src.llm import get_llm
    from langchain_core.messages import HumanMessage

    llm = get_llm(temperature=0)
    start = time.time()
    response = llm.invoke([HumanMessage(content="Say 'connected' and nothing else.")])
    elapsed = time.time() - start

    assert response is not None, "❌ No response from OpenRouter"
    assert "connected" in response.content.lower(), f"❌ Unexpected response: {response.content[:50]}"
    assert elapsed < 30, f"❌ Response took {elapsed:.0f}s (expected < 30s)"

    # Verify token usage in metadata
    usage = response.response_metadata.get("token_usage", {})
    assert usage.get("prompt_tokens", 0) > 0, "❌ No prompt tokens in response metadata"
    assert usage.get("completion_tokens", 0) > 0, "❌ No completion tokens in response metadata"

    print(f"  ✅ OpenRouter connected: {elapsed:.1f}s, "
          f"tokens={usage['prompt_tokens']}+{usage['completion_tokens']}")


# ─── Test 3: Structured Output ────────────────────────────────
def test_structured_output():
    """Verify with_structured_output returns a typed object."""
    from src.llm import get_llm
    from langchain_core.messages import HumanMessage, SystemMessage
    from pydantic import BaseModel, Field

    class SmokeTest(BaseModel):
        status: str = Field(description="Must be 'ok'")
        model_used: str = Field(description="The model that processed this request")

    llm = get_llm(temperature=0)
    structured = llm.with_structured_output(SmokeTest)

    response = structured.invoke([
        SystemMessage(content="Return valid JSON only."),
        HumanMessage(content="Set status='ok'. Include the model you're running on.")
    ])

    assert isinstance(response, SmokeTest), \
        f"❌ Expected SmokeTest, got {type(response).__name__}"
    assert response.status == "ok", f"❌ Expected status='ok', got '{response.status}'"
    assert response.model_used, "❌ model_used field is empty"

    print(f"  ✅ Structured output: {type(response).__name__}(status='{response.status}')")


# ─── Test 4: Cost Tracking ────────────────────────────────────
def test_cost_tracking():
    """Verify token usage and cost are available from response metadata."""
    from src.llm import get_llm
    from langchain_core.messages import HumanMessage

    llm = get_llm(temperature=0)
    response = llm.invoke([HumanMessage(content="One word: hello")])

    usage = response.response_metadata.get("token_usage", {})
    prompt_tokens = usage.get("prompt_tokens", 0)
    completion_tokens = usage.get("completion_tokens", 0)

    assert prompt_tokens > 0, "❌ prompt_tokens is 0"
    assert completion_tokens > 0, "❌ completion_tokens is 0"

    # Rough cost estimate
    cost = (prompt_tokens * 0.15 + completion_tokens * 0.60) / 1_000_000

    print(f"  ✅ Cost tracked: {prompt_tokens}+{completion_tokens} tokens, ~${cost:.6f}")


# ─── Test 5: Model Swap (Optional) ────────────────────────────
def test_model_swap():
    """Verify a different model also works (runs if DEVBUDDY_MODEL_ALT is set)."""
    if not DEVBUDDY_MODEL_ALT:
        print("  ⏭️  Model swap skipped (set DEVBUDDY_MODEL_ALT in .env to test)")
        return

    from src.llm import get_llm
    from langchain_core.messages import HumanMessage

    llm = get_llm(model=DEVBUDDY_MODEL_ALT, temperature=0)
    start = time.time()
    response = llm.invoke([HumanMessage(content="Say 'ok'")])
    elapsed = time.time() - start

    assert response is not None, f"❌ No response from {DEVBUDDY_MODEL_ALT}"
    assert elapsed < 30, f"❌ {DEVBUDDY_MODEL_ALT} took {elapsed:.0f}s"

    print(f"  ✅ Model swap: {DEVBUDDY_MODEL_ALT} responded in {elapsed:.1f}s")


# ─── Run All ──────────────────────────────────────────────────
def main():
    print("=" * 60)
    print("  DevBuddy Integration Test — Week 0")
    print("=" * 60)
    print()

    tests = [
        ("Environment", test_env_configured),
        ("OpenRouter Connectivity", test_openrouter_connectivity),
        ("Structured Output", test_structured_output),
        ("Cost Tracking", test_cost_tracking),
        ("Model Swap", test_model_swap),
    ]

    passed = 0
    failed = 0

    for name, test_fn in tests:
        try:
            test_fn()
            passed += 1
        except AssertionError as e:
            print(f"  {e}")
            failed += 1
        except Exception as e:
            print(f"  ❌ {name}: {e}")
            failed += 1

    print()
    print("=" * 60)
    if failed == 0:
        print(f"  ✅ ALL {passed} TESTS PASSED")
        print("  Your environment is fully configured and working.")
    else:
        print(f"  {passed} passed, {failed} failed")
        print("  Fix the failures above and re-run.")
    print("=" * 60)

    return failed == 0


if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: nodejs/tests/test_integration.js
/**
 * DevBuddy Integration Test — Week 0 (Node.js)
 *
 * Validates the full setup end-to-end:
 * 1. .env is configured (OPENROUTER_API_KEY, DEVBUDDY_MODEL)
 * 2. OpenRouter connectivity works
 * 3. Structured output returns typed objects
 * 4. Cost tracking works
 * 5. Model swap works (optional)
 *
 * Run: npm test
 */
import { describe, it, expect } from "vitest";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";
import { z } from "zod";
import { config, calculateCost } from "../src/config.js";
import { getLlm } from "../src/llm.js";

// ─── Test 1: Environment ──────────────────────────────────────
describe("Environment", () => {
  it("has OPENROUTER_API_KEY configured", () => {
    expect(config.openRouterApiKey).toBeTruthy();
    expect(config.openRouterApiKey).toContain("sk-or-");
  });

  it("has DEVBUDDY_MODEL configured", () => {
    expect(config.model).toBeTruthy();
  });
});

// ─── Test 2: OpenRouter Connectivity ──────────────────────────
describe("OpenRouter Connectivity", () => {
  it("responds to a basic prompt within 30s", async () => {
    const llm = getLlm();
    const start = performance.now();
    const response = await llm.invoke([
      new HumanMessage("Say 'connected' and nothing else."),
    ]);
    const elapsed = (performance.now() - start) / 1000;

    expect(response).toBeTruthy();
    expect(response.content.toLowerCase()).toContain("connected");
    expect(elapsed).toBeLessThan(30);

    // Verify token usage in metadata
    const usage = response.usage_metadata || {};
    const inputTokens = usage.input_tokens || 0;
    const outputTokens = usage.output_tokens || 0;

    console.log(
      `    ✅ OpenRouter connected: ${elapsed.toFixed(1)}s, ` +
        `tokens=${inputTokens}+${outputTokens}`
    );
  });
});

// ─── Test 3: Structured Output ────────────────────────────────
describe("Structured Output", () => {
  it("returns a typed Zod object", async () => {
    const SmokeTestSchema = z.object({
      status: z.string().describe("Must be 'ok'"),
      model_used: z
        .string()
        .describe("The model that processed this request"),
    });

    const llm = getLlm();
    const structured = llm.withStructuredOutput(SmokeTestSchema);

    const response = await structured.invoke([
      new SystemMessage("Return valid JSON only."),
      new HumanMessage(
        "Set status='ok'. Include the model you're running on."
      ),
    ]);

    expect(response).toBeTruthy();
    expect(response.status).toBe("ok");
    expect(response.model_used).toBeTruthy();

    console.log(
      `    ✅ Structured output: ${response.constructor.name}(status='${response.status}')`
    );
  });
});

// ─── Test 4: Cost Tracking ────────────────────────────────────
describe("Cost Tracking", () => {
  it("captures token usage on a plain LLM call", async () => {
    const llm = getLlm();
    const response = await llm.invoke([
      new HumanMessage("One word: hello"),
    ]);

    const usage = response.usage_metadata || {};
    const inputTokens = usage.input_tokens || 0;
    const outputTokens = usage.output_tokens || 0;

    expect(inputTokens).toBeGreaterThan(0);
    expect(outputTokens).toBeGreaterThan(0);

    // Verify our calculateCost helper works with these tokens
    const cost = calculateCost(inputTokens, outputTokens);
    expect(cost).toBeGreaterThan(0);

    console.log(
      `    ✅ Cost tracked: ${inputTokens}+${outputTokens} tokens, ` +
        `~$${cost.toFixed(6)}`
    );
  });
});

// ─── Test 5: Model Swap (Optional) ────────────────────────────
describe("Model Swap", () => {
  const altModel = config.modelAlt;

  if (altModel) {
    it(`swaps to ${altModel} successfully`, async () => {
      const llm = getLlm({ model: altModel });
      const start = performance.now();
      const response = await llm.invoke([
        new HumanMessage("Say 'ok'"),
      ]);
      const elapsed = (performance.now() - start) / 1000;

      expect(response).toBeTruthy();
      expect(elapsed).toBeLessThan(30);

      console.log(
        `    ✅ Model swap: ${altModel} responded in ${elapsed.toFixed(1)}s`
      );
    });
  } else {
    it.skip("model swap (set DEVBUDDY_MODEL_ALT in .env to enable)", () => {
      // Skipped — no alt model configured
    });
  }
});
